spring-web/src/main/java/org/springframework/http/RequestEntity.java

```diff
@@ -169,9 +195,27 @@ public HttpMethod getMethod() {
 	 * @return the URL as a {@code URI}
 	 */
 	public URI getUrl() {
-		return this.url;
+		if (uriTemplate == null) {
+			return this.url;
+		}
+
+		final DefaultUriBuilderFactory defaultUriBuilderFactory = new DefaultUriBuilderFactory();
+		defaultUriBuilderFactory.setEncodingMode(DefaultUriBuilderFactory.EncodingMode.URI_COMPONENT);
+		return defaultUriBuilderFactory.expand(uriTemplate.toString(), uriVariables);
```

Comment:
This factory be a default instance, not customized, and it should be a static variable so it is not re-created for every request.

Reply:
done